internal/db/users/user_repository.go

```diff
@@ -0,0 +1,133 @@
+	"go.mongodb.org/mongo-driver/mongo"
+
+	"github.com/TheDevExperiment/server/internal/db"
+	"github.com/TheDevExperiment/server/internal/db/users/models"
+	"github.com/TheDevExperiment/server/internal/log"
+	"github.com/TheDevExperiment/server/internal/utility/funnynamer"
+	"github.com/TheDevExperiment/server/internal/utility/jwt"
+)
+
+type UserRepository struct {
+	collection *mongo.Collection
+}
+
+var singletonUserCollection *mongo.Collection
+
+func NewUserRepository() *UserRepository {
+	if singletonUserCollection == nil {
+		log.Debug("Created Connection to user collection")
+		singletonUserCollection = db.GetCollection(viper.GetString("mongodb.db_name"), db.CollectionUsers)
+	}
+	return &UserRepository{collection: singletonUserCollection}
+}
+
+func (r *UserRepository) FindById(ctx context.Context, id string) (*models.User, error) {
+	bsonFilter := bson.M{db.FieldId: id}
+	oid, err := primitive.ObjectIDFromHex(id)
+	if err != nil {
+		return nil, err
+	}
+	bsonFilter[db.FieldId] = oid
+
+	result := r.collection.FindOne(ctx, bsonFilter)
+	if result.Err() != nil {
+		if result.Err() == mongo.ErrNoDocuments {
+			return nil, nil
+		}
+		return nil, result.Err()
+	}
+	var user models.User
+	if err := result.Decode(&user); err != nil {
+		return nil, err
+	}
+	return &user, nil
+}
+
+func (r *UserRepository) Create(ctx context.Context, userAge string, countryId string, cityId string) (models.User, error) {
+	doc := models.User{
+		IsGuest:         true,
+		GuestAuthSecret: "",
+		CreatedAt:       time.Now(),
+		LastModified:    time.Now(),
+		DisplayName:     funnynamer.GetFunnyName(),
+		CountryID:       countryId,
+		CityID:          cityId,
+		Age:             userAge,
+		Score:           models.Score{RatingScoreSum: 0, RatingCount: 0},
+		IsActive:        true,
+		DeletionReason:  "",
+	}
+
+	result, err := r.collection.InsertOne(ctx, doc)
+	if err != nil {
+		return models.User{}, err
+	}
+
+	insertedID, ok := result.InsertedID.(primitive.ObjectID)
+	if !ok {
+		return models.User{}, errors.New("failed to convert InsertedID to ObjectID")
+	}
+	secret, err := jwt.CreateToken(insertedID.Hex(), countryId, cityId, true)
+	if err != nil {
+		return models.User{}, err
+	}
+
+	deltaUpdate := models.UserUpdateModel{
+		GuestAuthSecret: secret,
+		LastModified:    time.Now(),
+	}
+	success, updateErr := r.UpdateById(ctx, insertedID.Hex(), deltaUpdate)
+	if updateErr != nil {
+		return models.User{}, updateErr
+	}
+	if !success {
+		return models.User{}, nil
+	}
+
+	doc.Id = insertedID.Hex()
+	doc.GuestAuthSecret = secret
+	doc.LastModified = time.Now()
+	doc.Score.RatingCount = 0
+	doc.Score.RatingScoreSum = 0
+	doc.IsActive = true
+	doc.DeletionReason = ""
+	return doc, nil
+}
+
+func (r *UserRepository) Delete(ctx context.Context, filter interface{}) error {
+	_, err := r.collection.DeleteMany(ctx, filter)
+	return err
+}
+
+func (r *UserRepository) UpdateById(ctx context.Context, id string, update models.UserUpdateModel) (bool, error) {
```

Comment:
I think the return params can be reduced to just `error`

true, nil -> nil
false, error -> error
false, nil -> this case can never happen

Reply:
i think **`false, nil`** can occur, although the possibility is less, but I've seen some example where modifiedCount is 0, even though query execution was successful (say filter is not correct logically, it will run fine but nothing will be updated)

Reply:
we should retain `false, nil` case incase you want to have some specific behaviour when that happens.. to keep things more generic, whenever the insertion fails due to filter error or whatever, you can return `fmt.Errorf("unsuccesful insert")`.. that way for callers, you wont always have to keep checks for all these cases, just one for `err!=nil`.. lemme know if you still feel have 2 returns are necessary here

Reply:
alright, make sense.
Modified